### DISCUSSION TITLE
Compile Nominatim without cloning osm2pgsql cloned and without PHP installed

### DISCUSSION THREAD

**Discussion by Yamakasi** [osm-search/Nominatim#2191] [Q&A] [answered]
Hi,

I want to offload PHP to other PHP servers so I like to compile Nominatim without PHP installed.

Is this possible ?

Also I like to not download osm2pgsql when cloning, I want to have only Nominatim to be compiled.

I assume that sharing the nominatim PHP files between my PHP backends and the Nominatim server(s) is enought to make this work ? Without having ngxin/apache local might be impossible ?

I run into these:

```
Cloning into 'Nominatim'...
Submodule 'osm2pgsql' (https://github.com/openstreetmap/osm2pgsql.git) registered for path 'osm2pgsql'
Cloning into '/opt/src/Nominatim/osm2pgsql'...
```

and:

```
CMake Warning at CMakeLists.txt:69 (message):
  pyosmium-get-changes not found (required for updates)


CMake Error at CMakeLists.txt:88 (message):
  PHP binary not found.  Install php or provide location with
  -DPHP_BIN=/path/php
```

Hoping that someone can clear this out.

Thanks!

**Comment by mtmail** [comment]
converting the issue to a discussion

**Comment by lonvia** [comment, marked as answer]
The cmake script defines various parameters to compile and install only parts of Nominatim:

```
BUILD_IMPORTER - Build everything for importing/updating the database
BUILD_API - Build everything for the API server
BUILD_MODULE - Build PostgreSQL module
BUILD_TESTS - Build test suite
BUILD_DOCS - Build documentation
BUILD_OSM2PGSQL - Build osm2pgsql
```

You can use those to switch off building parts of Nominatim. If you want to run the API on a different machine, you could switch off everything except BUILD_API. With `-DBUILD_OSM2PGSQL=off` the osm2pgsql submodule is not necessary. (Word of warning: these switches are rarely used and not well tested, so you might stumble over some bug you might want to report.)

You can ignore the non-fatal warnings about pyosmium.

**Reply by di-rect** [reply]
OK great, thanks for the reply!

You talk about API, I wonder... if I use a PHP library against Nominatim I don't need the PHP processing for that at all on the Nominatim side ? Because that would be awesome!

/edit oops... someone else was loggedin on this browser it seems :) Sorry!

**Reply by Yamakasi** [reply]
> The cmake script defines various parameters to compile and install only parts of Nominatim:
> 
> ```
> BUILD_IMPORTER - Build everything for importing/updating the database
> BUILD_API - Build everything for the API server
> BUILD_MODULE - Build PostgreSQL module
> BUILD_TESTS - Build test suite
> BUILD_DOCS - Build documentation
> BUILD_OSM2PGSQL - Build osm2pgsql
> ```
> 
> You can use those to switch off building parts of Nominatim. If you want to run the API on a different machine, you could switch off everything except BUILD_API. With `-DBUILD_OSM2PGSQL=off` the osm2pgsql submodule is not necessary. (Word of warning: these switches are rarely used and not well tested, so you might stumble over some bug you might want to report.)
> 
> You can ignore the non-fatal warnings about pyosmium.

I run into this:

```
-- Build files have been written to: /opt/src/build
Scanning dependencies of target nominatim_lib
[100%] Running external makefile /usr/lib/postgresql/pgxs/src/makefiles/pgxs.mk
/opt/src/Nominatim/module/Makefile:11: warning: overriding recipe for target 'install'
/usr/lib/postgresql/pgxs/src/makefiles/pgxs.mk:241: warning: ignoring old recipe for target 'install'
[100%] Built target nominatim_lib
make: *** No rule to make target 'install'.  Stop.
```

When using:

```
cmake -DBUILD_IMPORTER=off -DBUILD_API=on -DBUILD_MODULE=on -DBUILD_TESTS=on -DBUILD_DOCS=off 
 -DBUILD_OSM2PGSQL=off ../Nominatim
```
